client/extensions/woocommerce/woocommerce-services/components/shipping-classes-field/index.js

```diff
@@ -0,0 +1,84 @@
+/** @format */
+
+/**
+ * External dependencies
+ */
+import React from 'react';
+import PropTypes from 'prop-types';
+import { uniqBy, map } from 'lodash';
+
+/**
+ * Internal dependencies
+ */
+import FormFieldset from 'components/forms/form-fieldset';
+import FormLabel from 'components/forms/form-label';
+import TokenField from 'components/token-field';
+import FormSettingExplanation from 'components/forms/form-setting-explanation';
+
+export default class ShippingClassesField extends React.Component {
+	static propTypes = {
+		id: PropTypes.string.isRequired,
+		title: PropTypes.string,
+		description: PropTypes.string,
+		value: PropTypes.array.isRequired,
+		updateValue: PropTypes.func,
+		options: PropTypes.array,
+	};
+
+	render() {
+		const { id, title, description, value, placeholder, options } = this.props;
+
+		// If there are no shipping classes, no input for them is needed.
+		if ( false === options || 0 === options.length ) {
+			return null;
+		}
+
+		return (
+			<FormFieldset>
+				<FormLabel htmlFor={ id }>{ title }</FormLabel>
+				<TokenField
+					id={ id }
+					name={ id }
+					placeholder={ placeholder }
+					value={ map( value, this.getNameFromId ) }
```

Comment:
Isn't the PHP/plugin side of these changes expecting to use the shipping class ids, not names? (Looking at: https://github.com/Automattic/woocommerce-services/pull/1469/files#diff-5e282430ce2ce9e8ea6f132279193e7dR576)

It seems that all changes made to `suggestions` have had mirrored changes made to `value` as well - can `TokenField`s not store different values than they use for suggestions? Is that not what the `saveTransform` function is for?

Reply:
Correct, PHP (and pretty much everything outside of the `ShippingClassesField`) expects IDs.

The reason I am constantly using `getNameFromId` and `getIdFromName` is that `saveTransform` does not work the way you (and initially me) expect. I also thought that it works like a filter that gets called before the `onChange` prop, but that is not the case.

`saveTransform` is also called much earlier, for example when random text input is being converted to an actual token. If you remember, eariler I was using it to internally convert tokens to string IDs and then convert them back to texts in `displayTransform`. However, if I do that, I also need to use IDs as suggestions and then partial matching does not work.